### PR TITLE
[`LevelDbStore`] Updated features & Usability of Interop

### PR DESCRIPTION
**Pull request by cschuchardt88**
# Change Log
- Enabled all the features of `leveldb`
- New Classes, Memory management and interoperability.

## Type of change

- [x] Optimization (the change is only an optimization)
- [x] Style (the change is only a code style for better maintenance or standard purpose)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by Jim8y** [comment]
change like this better split into multiple small tasks and start working on a specific branch.

**Comment by cschuchardt88** [comment]
Why is so hard for you review a PR? No more than 10 lines, everyone throws a fit. It an application as a whole. No one is going to approve a PR if it doesn't work or incomplete application. It like me making a PR for a `HelloWorld` console app.  Even if it was on another branch it not going to get approve until years and could even be forgot. Do it right the 1st time and I wouldn't need to clean up. There is no thought put into code design. Not my fault and everyone wonders why no one uses it. So stop saying small PR. Small PRs are for bug fixes. How are we to expand? Plus I am moving this from `neo-modules`. My code is straight forward and easy to follow. Sorry for out burst but its at everyone not you @Jim8y.

**Comment by Jim8y** [comment]
i have being following with this issue for a while, and have discussed with multiple coredevs, turns out everyone is actually working hard and busy reviewing. the real problem i believe is the coredev system itself, reviewers are not responsible for reviewing a pr, and big prs usually will potentially have more issues and need more time to check, but with the same amount of time, we can revoew many other smaller  prs. its a system issue, not a single one of us can change it, but adapt it.

**Comment by Jim8y** [comment]
so i highly recommand you to split it. or i can do it for you without taking your bonus.

File: src/Plugins/LevelDBStore/Result.cs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+// accompanying file LICENSE in the main directory of the
+// repository or http://www.opensource.org/licenses/mit-license.php
+// for more details.
+//
+// Redistribution and use in source and binary forms with or without
+// modifications are permitted.
+
+namespace LevelDB
+{
+    //public class Result : LevelDBHandle, IEnumerable<byte>, IEnumerable<int>
```
Comment (shargon) [inline src/Plugins/LevelDBStore/Result.cs:14]: Remove this file

File: src/Plugins/LevelDBStore/SnapShot.cs [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    /// accessed from multiple threads without any external synchronization.
+    /// </summary>
+    public class SnapShot : LevelDBHandle
+    {
+        // pointer to parent so that we can call ReleaseSnapshot(this) when disposed
+        public WeakReference Parent;  // as DB
+
+        internal SnapShot(IntPtr Handle, DB parent)
+        {
+            this.Handle = Handle;
```
Comment (shargon) [inline src/Plugins/LevelDBStore/SnapShot.cs:27]:
```suggestion
        internal SnapShot(IntPtr handle, DB parent)
        {
            this.Handle = handle;
```

File: src/Plugins/LevelDBStore/SnapShot.cs [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+            Parent = new WeakReference(parent);
+        }
+
+        internal SnapShot(IntPtr Handle)
+        {
+            this.Handle = Handle;
```
Comment (shargon) [inline src/Plugins/LevelDBStore/SnapShot.cs:33]:
```suggestion
        internal SnapShot(IntPtr handle)
        {
            this.Handle = handle;
```

File: src/Plugins/LevelDBStore/Iterator.cs [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+namespace LevelDB
+{
+    /// <summary>
+    /// An iterator yields a sequence of key/value pairs from a database.
+    /// </summary>
+    public class Iterator : LevelDBHandle
+    {
+        private readonly Encoding _encoding;
+
+        internal Iterator(IntPtr Handle, Encoding encoding)
```
Comment (shargon) [inline src/Plugins/LevelDBStore/Iterator.cs:24]: here also we have args in uppercase

**Comment by cschuchardt88** [comment]
I could see this being a problem. If we have deadlines. But we don't. Its only `32` files, many small. It show be the job of the tester, to test it and use it for some time, to make sure its in working order. Also the job of the developer doing the PR to do the testing. Most PR's are un-tested before merge. So I can understand your point of view. Just know this I test and test my code before I make PR. I am over compulsive about style, code functionality and code design.

File: src/Plugins/LevelDBStore/Logger.cs [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+namespace LevelDB
+{
+    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
+    public delegate void Log(string msg);
+
+    public class Logger : LevelDBHandle
+    {
+        public Logger(Log log)
+        {
+            var p = Marshal.GetFunctionPointerForDelegate(log);
```
Comment (shargon) [inline src/Plugins/LevelDBStore/Logger.cs:24]: Could you add a test for log?

File: src/Plugins/LevelDBStore/Options.cs [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+        ///    ~400-800MB/s decompression
+        /// Note that these speeds are significantly faster than most
+        /// persistent storage speeds, and therefore it is typically never
+        /// worth switching to kNoCompression.  Even if the input data is
+        /// incompressible, the kSnappyCompression implementation will
+        /// efficiently detect that and will switch to uncompressed mode.
+        /// </summary>
+        public CompressionLevel CompressionLevel
+        {
+            set { LevelDBInterop.leveldb_options_set_compression(Handle, (int)value); }
```
Comment (shargon) [inline src/Plugins/LevelDBStore/Options.cs:189]: I didn't like only set properties, and I think that I never used a popular library with these kind of properties

File: src/Plugins/LevelDBStore/Options.cs [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+
+namespace LevelDB
+{
+    /// <summary>
+    /// Options to control the behavior of a database (passed to Open)
+    /// 
+    /// the setter methods for InfoLogger, Env, and Cache only "safe to clean up guarantee". Do not
+    /// use Option object if throws.
+    /// </summary>
+    public class Options : LevelDBHandle
```
Comment (shargon) [inline src/Plugins/LevelDBStore/Options.cs:20]: All to `internal`?

**Comment by lingyido** [comment]
Please take all your time for this one. I see no good reason for updating new features for levelDB which is an essential modules. I personally disagree these great changes.

**Comment by cschuchardt88** [comment]
> Please take all your time for this one. I see no good reason for updating new features for levelDB which is an essential modules. I personally disagree these great changes.

Its more stable and done properly in dotnet for the GC and other things.

File: src/Plugins/LevelDBStore/Logger.cs (hunk 5 above)
Comment (cschuchardt88) [reply src/Plugins/LevelDBStore/Logger.cs:24]: Sure

File: src/Plugins/LevelDBStore/Options.cs (hunk 6 above)
Comment (cschuchardt88) [reply src/Plugins/LevelDBStore/Options.cs:189]: You can set update the properties at anytime. Which is why its like this.

File: src/Plugins/LevelDBStore/Options.cs (hunk 7 above)
Comment (cschuchardt88) [reply src/Plugins/LevelDBStore/Options.cs:20]: We could but its a plugin.

Comment (cschuchardt88) [reply src/Plugins/LevelDBStore/Options.cs:20]: I thought about it. This is a `plugin` and its best to have it be `public`.

**Comment by cschuchardt88** [comment]
### No Performance Increase
image

**Comment by cschuchardt88** [comment]
Check Simplified PR #3414 with minimal changes

**Pull request closed without merging**